File: go/medium/add_two_numbers.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	n := ListNode{}
	m := &n
	a := 0
	for l1 != nil || l2 != nil || a != 0 {
		s := a
		if l1 != nil {
			s = s + l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			s = s + l2.Val
			l2 = l2.Next
		}
		a = s / 10
		m.Next = &ListNode{ Val: s % 10 }
        m = m.Next
	}
	return n.Next
}

File: javascript/medium/add_two_numbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
    let n = new ListNode(0);
    let m = n;
    let a = 0;
    while (l1 || l2 || a !== 0) {
        let s = a;
        if (l1) {
            s = s + l1.val;
            l1 = l1.next;
        }
        if (l2) {
            s = s + l2.val;
            l2 = l2.next;
        }
        a = Math.floor(s / 10);
        m.next = new ListNode(s % 10);
        m = m.next;
    }
    return n.next;
};
